syntax = "proto3";

package protocir;
option java_package = "org.jacodb.impl.grpc";

message CIRModuleID { string id = 1; }

message CIRGlobalID {
  CIRModuleID module_id = 1;
  string id = 2;
}

message CIRFunctionID {
  CIRModuleID module_id = 1;
  string id = 2;
}

message CIRBlockID { uint64 id = 1; }

message CIRTypeID {
  CIRModuleID module_id = 1;
  string id = 2;
}

message CIROpID { uint64 id = 1; }

message CIRValue {
  oneof value {
    CIROpValue op_value = 1;
    CIRArgValue arg_value = 2;
  }
}

message CIROpValue {
  CIROpID op_id = 1;
  uint64 result_id = 2;
}

message CIRArgValue {
  CIRBlockID block_id = 1;
  uint64 arg_id = 2;
}

message CIRValues { repeated CIRValue range = 1; }
